canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -131,57 +133,138 @@ private[apiserver] final class ApiPackageManagementService private (
     } yield dar
   }
 
-  private def validateUpgrade(upgradingPackage: Dar[(Ref.PackageId, Ast.Package)])(implicit
-      loggingContext: LoggingContextWithTrace
+  private def lookupDar(pkgId: Ref.PackageId)(implicit
+                                                  loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    for {
+      optArchive
+        <- packagesIndex.getLfArchive(pkgId)
+      optPackage
+        <- Future.fromTry {
+        optArchive.traverse(Decode.decodeArchive(_))
+          .handleError(Validation.handleLfArchiveError)
+      }
+    } yield optPackage
+  }
+
+  private def existingVersionedPackageId(dar: (Ref.PackageId, Ast.Package)): Option[Ref.PackageId] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.name
+    val pkgVersion = pkg.metadata.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collectFirst { case (pkgId, (`pkgName`, `pkgVersion`)) => pkgId }
+  }
+
+  private def minimalVersionedDar(upgradedPackageId: Ref.PackageId, dar: (Ref.PackageId, Ast.Package))(implicit
+                                                                loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.name
+    val pkgVersion = pkg.metadata.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collect { case (pkgId, (`pkgName`, pkgVersion)) =>
+      (pkgId, pkgVersion)
+    }
+      .filter { case (pkgId, version) => pkgVersion < version && upgradedPackageId != pkgId }
+      .minByOption { case (_, version) => version }
+      .traverse { case (pId, _) => lookupDar(pId) }
+      .map(_.flatten)
+  }
+
+  private def maximalVersionedDar(upgradedPackageId: Ref.PackageId, dar: (Ref.PackageId, Ast.Package))(implicit
+                                                                loggingContext: LoggingContextWithTrace
+  ): Future[Option[(Ref.PackageId, Ast.Package)]] = {
+    val (_, pkg) = dar
+    val pkgName = pkg.metadata.name
+    val pkgVersion = pkg.metadata.version
+    packageMetadataStore.getSnapshot.getUpgradablePackageMap.collect { case (pkgId, (`pkgName`, pkgVersion)) =>
+      (pkgId, pkgVersion)
+    }
+      .filter { case (pkgId, version) => pkgVersion > version && upgradedPackageId != pkgId }
+      .maxByOption { case (_, version) => version }
+      .traverse { case (pId, _) => lookupDar(pId) }
+      .map(_.flatten)
+  }
```

Comment:
Is this identical to `minimalVersionedDar` with the exception of the comparison between `pkgVersion` and `version`? I don't have a strong preference here, but I'd personally deduplicate the code and maybe pass a the comparison as an argument (e.g. `_ < _`). Feel free to ignore if you prefer it like this

Reply:
Alright 👍